pkg/services/pluginsintegration/cachekvstore/namespacedstore.go

```diff
@@ -0,0 +1,166 @@
+	// This allows to modify the key for the underlying storage.
+	// This is only used for actual data, not for the last updated time key.
+	storeKeyGetter StoreKeyGetter
+
+	// lastUpdatedKey is the key to use for the last updated time key.
+	lastUpdatedKey string
+
+	// setLastUpdatedOnDelete is a flag that determines if the last updated time should be updated when deleting a key.
+	setLastUpdatedOnDelete bool
+}
+
+// DefaultStoreKeyGetter is the default StoreKeyGetterFunc, which returns the key as-is.
+var DefaultStoreKeyGetter = StoreKeyGetterFunc(func(k string) string {
+	return k
+})
+
+// PrefixStoreKeyGetter returns a StoreKeyGetterFunc that returns the key with a prefix.
+// It is used to prefix all keys in the underlying store with a fixed prefix.
+func PrefixStoreKeyGetter(prefix string) StoreKeyGetterFunc {
+	return func(k string) string {
+		return prefix + k
+	}
+}
+
+// NamespacedStoreOpt is an option for NewNamespacedStore that modifies the store passed to it.
+type NamespacedStoreOpt func(store *NamespacedStore)
+
+// WithLastUpdatedKey sets the key to use for the last updated time key.
+func WithLastUpdatedKey(lastUpdatedKey string) NamespacedStoreOpt {
+	return func(store *NamespacedStore) {
+		store.lastUpdatedKey = lastUpdatedKey
+	}
+}
+
+// WithStoreKeyGetter sets the StoreKeyGetter to use.
+func WithStoreKeyGetter(g StoreKeyGetter) NamespacedStoreOpt {
+	return func(store *NamespacedStore) {
+		store.storeKeyGetter = g
+	}
+}
+
+// WithSetLastUpdatedOnDelete sets the setLastUpdatedOnDelete flag.
+func WithSetLastUpdatedOnDelete(updateLastUpdatedOnDelete bool) NamespacedStoreOpt {
+	return func(store *NamespacedStore) {
+		store.setLastUpdatedOnDelete = updateLastUpdatedOnDelete
+	}
+}
+
+// NewNamespacedStore creates a new NamespacedStore using the provided underlying KVStore and namespace.
+func NewNamespacedStore(kv kvstore.KVStore, namespace string, opts ...NamespacedStoreOpt) *NamespacedStore {
+	store := &NamespacedStore{
+		kv:                     kvstore.WithNamespace(kv, 0, namespace),
+		storeKeyGetter:         DefaultStoreKeyGetter,
+		lastUpdatedKey:         DefaultLastUpdatedKey,
+		setLastUpdatedOnDelete: true,
+	}
+	for _, opt := range opts {
+		opt(store)
+	}
+	return store
+}
+
+// Get returns the value for the given key.
+// If no value is present, the second argument is false and the returned error is nil.
+func (s *NamespacedStore) Get(ctx context.Context, key string) (string, bool, error) {
+	return s.kv.Get(ctx, s.storeKeyGetter.GetStoreKey(key))
+}
+
+// Set sets the value for the given key and updates the last updated time.
+// It uses the Marshal method to marshal the value before storing it.
+// This means that the value to store can implement the Marshaler interface to control how it is stored.
+func (s *NamespacedStore) Set(ctx context.Context, key string, value any) error {
+	valueToStore, err := Marshal(value)
```

Comment:
I think that you can omit all the `Marshaler` implementation if you just do the `Marshal` function here. You don't need to expose either:

```go
func marshal(value any) (string, error)
```

I don't think the customization for this Marshal method is needed.